Dev/fronthaul manager

morantr opened the pull request:
This PR creates a new process on the Agent - "Fronthaul_manager", and moves the ap_manager and the monitor threads to it.

See phase 1 on #1082

I am sorry it is a BIG pull request with a lot of commits, but all of them are needed to be in single PR since without all of them the code is broken or not passing the GitHub checks.

vitalybu commented on framework/external/easylogging/easylogging++.cc:
```diff
@@ -2605,7 +2605,12 @@ Writer& Writer::construct(int count, const char* loggerIds, ...) {
+      m_logger = ELPP->registeredLoggers()->get(loggerId, ELPP->hasFlag(LoggingFlag::CreateLoggerAutomatically));
+    } else {
+      m_logger = ELPP->registeredLoggers()->get(thread_logger_id, ELPP->hasFlag(LoggingFlag::CreateLoggerAutomatically));
+    }
```
```suggestion
    auto logger_id = ELPP->getThreadLoggerId().empty() ? loggerId : ELPP->getThreadLoggerId();
    m_logger = ELPP->registeredLoggers()->get(logger_id, ELPP->hasFlag(LoggingFlag::CreateLoggerAutomatically));
```

vitalybu replied:
But, I don't think this change should be here.
You should pass the correct `loggerId` in the first place (from whoever initializes the `Writer` [constructor?]).
When is the `Writer` initialized anyway?

morantr replied:
I don't like your suggestion since defining a new variable with the same name as one of the function's argument is confusing. Also, the cppcheck will complain about it that it is a variable shadowing.

A `Writer` is an object that is constructed when calling to `LOG(LEVEL)` or `CLOG(LEVEL, ID)`.

> You should pass the correct loggerId in the first place (from whoever initializes the Writer [constructor?]).

Yes, but we don't have it on external libs, which is the whole point of this commit.

arnout replied:
> I don't like your suggestion since defining a new variable with the same name as one of the function's argument is confusing. Also, the cppcheck will complain about it that it is a variable shadowing.

I @vitalybu's point was not about the variable name, but rather about limiting the condition to the part which is different (i.e. not repeat the `ELPP->registeredLoggers()->get(foo, ELPP->hasFlag(LoggingFlag::CreateLoggerAutomatically))` part). Of course, by all means, choose a better variable name.

vitalybu replied:
So change the name of the variable.. It's not the point here.
So do you mean we don't have it on external libs? What external libs?

morantr replied:
External lib is not accurate. Just "lib".
Rember we talked about that If we want to write to specific logger ID we have to use `CLOG(LEVEL, ID)`.
So we did macro hack for that that could work on files of a specific process, but we had a problem with a print in a lib.
For example if there is a function on network_utils that print something. So in order to make it print to the right file, we would have to provide the longer ID to each function and change every LOG(LEVEL) to CLOG(LEVEL, ID).
So this is the solution that allows us not to provide this, and instead identifying it automatically.

Anyway, I narrowed the condition as you wanted.

vitalybu commented on framework/external/easylogging/easylogging++.h:
```diff
@@ -2694,6 +2694,22 @@ class Storage : base::NoCopy, public base::threading::ThreadSafe {
+    if (found == m_threadIdLoggerId.end()) {
+      return std::string();
+    }
+    return m_threadIdLoggerId.at(base::threading::getCurrentThreadId());
```
```suggestion
    return (found == m_threadIdLoggerId.end()) ? std::string() : found.second;
```

morantr replied:
added fixup

vitalybu commented on agent/src/beerocks/fronthaul_manager/monitor/monitor_thread.cpp:
```diff
@@ -213,6 +213,11 @@ bool monitor_thread::init()
 
 void monitor_thread::after_select(bool timeout)
 {
+    if (!logger_thread_name_set) {
```
`logger_thread_name_set` - name?
Also should probably be moved to `before_select()`

morantr replied:
Oops! While I wrote it I set the thread on the logger but then I removed it and forgot to change it. I will rename this flag. I also forgot "m_" prefix.
Anyway, let me check it with syslog first, maybe I would have to set the thread name again for it.

The monitor thread does not override `before_select()`. It is redundant to add it just for that.

vitalybu replied:
It's perfectly fine to add it just for that.
You want to set apply this configuration as soon as possible.

vitalybu commented on framework/external/easylogging/easylogging++.h:
```diff
@@ -3688,6 +3706,14 @@ class Helpers : base::StaticClass {
+  static inline void attachLoggerIdThreadId(const std::string& logger_id) {
+    ELPP->attachLoggerIdThreadId(logger_id);
+  }
+  static inline const std::string getThreadLoggerId() {
```
```suggestion
  static inline const std::string getCurrentThreadLoggerId() {
```

morantr replied:
added fixup

vitalybu commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -17,6 +17,8 @@ namespace beerocks {
 #define BEEROCKS_CONTROLLER "beerocks_controller"
 #define BEEROCKS_AGENT "beerocks_agent"
+#define BEEROCKS_FRONTHAUL "beerocks_fronthaul"
+#define BEEROCKS_AP_MANAGER "beerocks_ap_manager"
```
This will be removed in the future?

morantr replied:
No. The BEEROCKS_FRONTHAUL is for pid file touch, and BEEROCKS_AP_MANAGER/MONITOR is for the logger.

arnout commented on common/beerocks/bcl/include/bcl/beerocks_logging.h:
```diff
@@ -73,12 +69,12 @@ class logging {
-                     const std::string &module_name = BEEROCKS_LOGGING_MODULE_NAME);
+    logging(const std::string &module_name, const std::string &config_path = std::string(),
+            const std::string &logger_id = std::string());
+    explicit logging(const std::string &module_name, const settings_t &settings,
```
There's no need for an explicit here any more, since it has two mandatory arguments.

However, now we need an explicit in the other constructor, since it has only a single argument so it would be called as an implicit conversion from string.

morantr replied:
Done.

arnout commented on common/beerocks/bcl/include/bcl/beerocks_logging.h:
```diff
@@ -129,6 +127,8 @@ class logging {
 
     settings_t m_settings_map;
+
+    std::string m_logger_id = "default";
```
(nitpick) since there is no default constructor and all constructors have a logger_id argument, I don't think there should be a default here.

morantr replied:
The logger ID is an optional argument. That's why it is needed.

arnout commented on common/beerocks/bcl/source/beerocks_logging.cpp:
```diff
@@ -301,13 +302,17 @@ logging::logging(const std::string &config_path, const std::string &module_name)
     if (!found_settings) {
         found_settings = load_settings(conf_path);
     }
+    if (!logger_id.empty()) {
+        m_logger_id = logger_id;
+    }
```
(nitpick) I think it would be better to put this in the initializer list, like `m_logger_id(logger_id.empty() ? "default" : logger_id)`

(same for the other constructors).

morantr replied:
But it is already "default" by default on the class definition. :)
Yes it will be one line, but it will increase the code size unnecessarily. So no.

arnout commented on common/beerocks/bcl/include/bcl/beerocks_logging.h:
```diff
@@ -73,12 +69,12 @@ class logging {
+            const std::string &logger_id = std::string());
+    explicit logging(const std::string &module_name, const settings_t &settings,
+                     const std::string &logger_id = std::string(), bool cache_settings = false);
     logging(const beerocks::config_file::SConfigLog &settings, const std::string &module_name,
```
Since module_name is the first argument in the other constructors, it would be better to do the same here.

morantr replied:
Done.

arnout commented on agent/src/beerocks/fronthaul_manager/fronthaul_manager_main.cpp:
```diff
@@ -48,6 +48,7 @@ static void handle_signal()
 
         s_pLogger->apply_settings();
         LOG(INFO) << "--- Start of file after roll ---";
+
```
(nitpick) unrelated and doesn't seem useful.

morantr replied:
Why not? It helps to understand when rolling has occurred.

arnout commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -91,12 +91,13 @@ static void copy_vaps_info(std::shared_ptr<bwl::ap_wlan_hal> &ap_wlan_hal,
 {
     thread_name = "ap_manager";
     slave_uds   = slave_uds_;
+    m_iface     = iface;
```
Why is this done with assignment instead of with initializer? ([C.49](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c49-prefer-initialization-to-assignment-in-constructors))

morantr replied:
I just hate to initialize things on the initializer list, and only do it for things I have to.
IMO it is ugly, and it very annoying that the compiler force you to initialize the members in the order they are in the class!
I would have done it if it is a struct that is often being created. But for a class that is created once, I prefer doing it that way.

arnout replied:
It seems that [the rest of the developers don't share your opinion](https://www.strawpoll.me/20160132).

That said, no need to change this right now. We have this pattern all over already at the moment. But keep it in mind for future changes.

morantr replied:
But "Preferer assignment in constructors" has the majority of votes.

arnout replied:
At the time I made the comment, only 9 people had voted - I thought that was about everybody but apparently not. Now 7 more have voted and they all chose the "prefer assignment in constructors" option.

arnout commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.h:
```diff
@@ -22,7 +24,8 @@ namespace son {
 
 public:
-    explicit ap_manager_thread(std::string slave_uds_);
+    explicit ap_manager_thread(const std::string &slave_uds_, const std::string &iface,
```
Explicit is no longer needed.

morantr replied:
Done.

arnout commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.h:
```diff
@@ -77,12 +80,19 @@ class ap_manager_thread : public beerocks::socket_thread {
     void allow_expired_clients();
 
+    // Class constants
+    static constexpr uint8_t BSS_STEER_IMMINENT_VALID_INT = 100;
```
Genuine question: why constexpr and not just const?

morantr replied:
constexpr is set on compile time. It is more like #define macro.
Honestly, I think we should forbid defining constants with `#define` and replace it with `constexpr`.

arnout commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -92,8 +92,6 @@ enum eGlobals {
     PHY_RATE_100KB_INVALID                  = 0,
     BSS_STEER_DISASSOC_TIMER_MS             = 200, // ~200ms
-    BSS_STEER_IMMINENT_VALID_INT            = 100,
-    BSS_STEER_VALID_INT                     = 50,
```
I like this!

tomereli commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -91,12 +91,13 @@ static void copy_vaps_info(std::shared_ptr<bwl::ap_wlan_hal> &ap_wlan_hal,
 {
     thread_name = "ap_manager";
     slave_uds   = slave_uds_;
+    m_iface     = iface;
```
If you add a member with m_, you're breaking the style in this file. I suggest to add a preparative commit which adds the _m to all members in this file first.

morantr replied:
I followed our coding style document. I can refactor this if you want.
But anyway it will be a separate commit at the end.
Please reply if you want me to add a commit that refactors this.

arnout replied:
Although I absolutely applaud the concept of doing small refactorings like this as part of a feature-oriented PR, it is not a good idea to do that if the PR already has 38 commits... So no, please don't do that as part of this PR.

morantr replied:
Ok.

tomereli commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.h:
```diff
@@ -77,12 +80,19 @@ class ap_manager_thread : public beerocks::socket_thread {
     void allow_expired_clients();
 
+    // Class constants
+    static constexpr uint8_t BSS_STEER_IMMINENT_VALID_INT = 100;
```
Why did you move these from beerocks_defines.h?

morantr replied:
Because it is only used in the ap_manager. There isn't a good reason for that to be on beerocks_defines.h.

tomereli commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -167,36 +167,6 @@ void ap_manager_thread::after_select(bool timeout)
 
     auto now = std::chrono::steady_clock::now();
 
```
Great commit and commit message!

morantr replied:
Thanks.

tomereli commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.h:
```diff
@@ -35,6 +35,8 @@ class ap_manager_thread : public beerocks::socket_thread {
         bool iface_filter_low;
     };
 
+    enum class eApManagerState { INIT, WAIT_FOR_CONFIGURATION, ATTACHING, ATTACHED, OPERATIONAL };
```
Please add a doxygen comment explaining each state.

morantr replied:
I think that the state name explains itself, and that trying to explain it will look ridiculous since I would have explain it with the state name. What's done in each state can be understood by looking at what is being done on the fsm function.

All that said, I think that documenting an enum/struct is weird and we should not do it.
We need to document what is not obvious like functions and logic. Documenting any other thing is redundant.

vitalybu replied:
Maybe in this case the cases are self-explanatory, but in other places this may not be the case.
Documentation is not meant for the developer writing it, but to others that may read the code (or even the generated doxygen file, and in this case they don't even have the source code to look at).

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:
```diff
@@ -202,10 +202,12 @@ eActionOp_APMANAGER:
-  ACTION_APMANAGER_ENABLE_APS_REQUEST: 2
-  ACTION_APMANAGER_ENABLE_APS_RESPONSE: 3
-  ACTION_APMANAGER_ACK: 4
+  ACTION_APMANAGER_UP_NOTIFICATION: 1
```
```suggestion
  ACTION_APMANAGER_IFACE_UP_NOTIFICATION: 1
```

morantr replied:
This message doesn't notify about an interface that went up, it notifies about the ap_manager thread is up and running.

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:
```diff
@@ -202,10 +202,12 @@ eActionOp_APMANAGER:
-  ACTION_APMANAGER_ENABLE_APS_RESPONSE: 3
-  ACTION_APMANAGER_ACK: 4
+  ACTION_APMANAGER_UP_NOTIFICATION: 1
+  ACTION_APMANAGER_CONFIGURE: 2
```
This one definitely needs a better name, I just don't know which one since I have no clue what configure means in this regard.

morantr replied:
It configures the following:
  iface_type
  channel
  iface_filter_low

These are all needed to configure the ap_hal.
The name seems ok to me. One can always look at which data is being sent on this message on the class implementation.

morantr replied:
Edit: I removed `iface_type` and `iface_filter_low` after @arnout comment.
https://github.com/prplfoundation/prplMesh/pull/1271#discussion_r426739414

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml:
```diff
@@ -16,6 +16,21 @@ _multi_class_auto_insert:
+  _type: class
+  iface_name_length:
+    _type: uint8_t
+    _length_var: True
```
I think that `_length_var` is used only for TLV classes, tlvf is using it to know if it needs to update the length var.
Therefore:
```suggestion
    iface_name_length: uint8_t
```

morantr replied:
It is a class, so it is ok :)

tomereli commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -132,30 +132,26 @@ bool ap_manager_thread::init()
+void ap_manager_thread::connect_to_agent()
+{
+    if (slave_socket) {
+        LOG(WARNING) << "Agent socket is already open";
```
IMO if slave socket is not nullptr, we should add the log but still restart the connection.

tomereli replied:
I now see you only call this when slave_socket is `nullptr`. In this case, this would be more appropriate:
`LOG_IF(slave_socket, FATAL) << "Agent socket is already open!"`

morantr replied:
FATAL??? It is a little bit exaggerated don't you think?
FATAL is for serious system errors. This is just a wrong use of the function and therefore prints warning and returns with doing nothing.

tomereli commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -158,9 +158,34 @@ void ap_manager_thread::ap_manager_fsm()
     switch (m_state) {
     case eApManagerState::INIT: {
+        if (!slave_socket) {
+            return;
```
Shouldn't there be a warning log here?

morantr replied:
Not critical. If there is an error when opening the socket, it will be printed on the connect function.

tomereli commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -158,9 +158,34 @@ void ap_manager_thread::ap_manager_fsm()
     }
     case eApManagerState::WAIT_FOR_CONFIGURATION: {
+        // On ACTION_APMANAGER_CONFIGURE handler, the ap_hal will be created, and the state will
+        // change to ATTACHING state.
```
Why do we need the "attaching" state? Can't we wait till attached or timeout? 
The name of the state is also weird - "WAIT_FOR_CONFIGURATION" - what configuration are we waiting here?
Perhaps a better name can be "WAIT_FOR_BWL_ATTACH" - we are waiting for the BWL to complete the attach FSM. Once it is done, we exit this state. If there is an error, we will be notified, or a timeout can occur. So I think we can save a couple of states here. But perhaps I'm missing something...

morantr replied:
> Why do we need the "attaching" state? Can't we wait till attached or timeout?

The attaching state is needed because the HAL fsm needs to from the ap_manager to call its "attach()" function until it succeeds to attach. This is how it worked before, the only difference it is now on fsm.

>The name of the state is also weird - "WAIT_FOR_CONFIGURATION" - what configuration are we waiting here?
Perhaps a better name can be "WAIT_FOR_BWL_ATTACH" - we are waiting for the BWL to complete the attach FSM. Once it is done, we exit this state. If there is an error, we will be notified, or a timeout can occur. So I think we can save a couple of states here. But perhaps I'm missing something...

Configuration message. I send it on few commits ahead. I mentioned it on the comment in the state.
The "WAIT_FOR_BWL_ATTACH" is not fit since it is just not how the fsm of the hal is working.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.h:
```diff
@@ -143,8 +142,8 @@ class slave_thread : public beerocks::socket_thread {
     bool hostap_services_on();
-    bool ap_manager_start();
-    void ap_manager_stop();
+    bool ap_manager_start() { return true; };
```
It looks a bit weird that these are nops.

morantr replied:
This is temporary so everything will be compiled successfully. It is removed few on a commit ahead.

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:
```diff
@@ -238,21 +238,18 @@ eActionOp_APMANAGER:
   ACTION_APMANAGER_CLIENT_BSS_STEER_REQUEST: 48
   ACTION_APMANAGER_CLIENT_BSS_STEER_RESPONSE: 49
   ACTION_APMANAGER_CLIENT_RX_RSSI_MEASUREMENT_CMD_RESPONSE: 50
```
I guess this should be part of the commit which removed the 4addr connection flow (nitpick)

morantr replied:
I preferred to split commit that is related to tlvf to make the review more convenient.

tomereli commented on agent/src/beerocks/fronthaul_manager/fronthaul_manager_main.cpp:
```diff
@@ -203,34 +204,52 @@ int main(int argc, char *argv[])
+            break;
+        }
+
+        // The ap_manager is the main process thread. After the ap_manager is finish the attach
```
```suggestion
        // The ap_manager is the main process thread. After the ap_manager finishes the attach
```

morantr replied:
actually it should be "is finished".

tomereli commented on agent/src/beerocks/slave/son_slave_thread.h:
```diff
@@ -86,18 +86,14 @@ class slave_thread : public beerocks::socket_thread {
         STATE_WAIT_FOR_BACKHAUL_MANAGER_CONNECTED_NOTIFICATION,
         STATE_BACKHAUL_MANAGER_CONNECTED,
         STATE_WAIT_BACKHAUL_MANAGER_BUSY,
+        STATE_WAIT_BEFORE_JOIN_MASTER,
```
Shouldn't this be explained in a bit more details (and perhaps as a separate commit)?

morantr replied:
Not critical, I've just moved it to the right location on the enum. Nothing new has been added.

tomereli commented on common/beerocks/bcl/include/bcl/beerocks_socket_thread.h:
```diff
@@ -97,9 +97,10 @@ class socket_thread : public thread_base {
+    uint32_t m_select_timeout_msec  = 0;
+    bool m_skip_next_select_timeout = false;
+    std::chrono::steady_clock::time_point m_select_wake_up_time =
+        std::chrono::steady_clock::time_point::min();
```
I'm not sure I understand the problem and fix, but this feels like a bit of a hack

morantr replied:
It is not a hack, it is a more correct initialization value. I'll try to explain.
There is code on the work() that check how much time it took the thread to do a single select cycle and print a warning message if it exceeds the select timeout value.
In the fronthaul, we create the monitor thread sooner, but we don't start the thread right away. So on the first cycle of the work function it will print a warning message.
This happens since I have chosen to use `now()`, and we get delta time which exceeds the select_timeout.
Choosing `min()` cause to the delta to be negative and thus the condition for select_timeout result in "false" and the error is not printed.

tomereli commented on tools/docker/static-analysis/cppcheck.sh:
```diff
@@ -21,6 +21,7 @@ run_cppcheck() {
              --force \
              --error-exitcode=1 \
              --enable=warning,style,information \
+             --suppressions-list="$rootdir"/tools/docker/static-analysis/suppressions.txt \
```
I don't understand this commit. Please add explanation.

morantr replied:
Add detailed commit message.

arnout commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -132,30 +132,26 @@ bool ap_manager_thread::init()
     }
 
-    //connect to slave //
+    return true;
```
I don't understand from the commit message why it is better to do the connect to agent from after_select instead of directly from init.

morantr replied:
If I remember correctly, it was because I planned to not call the init function from the fronthaul main file.
The only thing left there is  ap_hal command that calls to a shell command "iwpriv" which I'm not sure is still working.

Anyway, I didn't check it and left the init function for now.

arnout commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -357,6 +354,32 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
+
+        low_filter   = config->iface_filter_low();
+        m_iface_type = config->iface_type();
+        wifi_channel = config->channel();
```
AFAICS, this is not used at all, so maybe remove it entirely?

Actually, originally it was used to set acs_enabled, but I don't see that any more? And also I think acs_enabled could be handled better than through the configuration. AFAICS it is only used in two places:
- in dwpal's attach function, but only for error messages;
- in handle_hostap_attached(), to choose between read_acs_report or read_preferred_channels.

Note that *doesn't* seem to be used to actually configure bwl into ACS, so I guess there's a paralle configuration knob in the wireless UCI config to enable ACS.

Therefore, it would make a whole lot more sense to me to handle the ACS option completely within bwl. Indeed, whether ACS is enabled or not, we should simply be able to read the channel preferences from hostapd - the difference is that hostapd will return different preference values.

Obviously, that's too big to include in this PR though...

morantr replied:
You are right, I didn't notice that they are not being used anymore.
I will remove `low_filter` and `wifi_channel`.
`m_iface_type` is actually do needed, and it should be set on base_wlan_hal() constructor. For some reason, it uses everywhere `IfaceType::Intel` although it could be another platform.
On the other hand, why we need to set it from outside so we could use get_iface_type() and send it back to ap_manager when son_slave already has this information.

I will remove the above from the config message and from the class. But I'll leave the `acs_enable`. Also, I've just noticed that I forgot to set this member.

arnout commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -357,6 +354,32 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
+        }
+
+        low_filter   = config->iface_filter_low();
+        m_iface_type = config->iface_type();
```
This doesn't seem to be used anywhere, maybe remove it?

morantr replied:
Answered on: https://github.com/prplfoundation/prplMesh/pull/1271#discussion_r426739414

arnout commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -339,6 +358,32 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
+            return false;
+        }
+
+        low_filter   = config->iface_filter_low();
```
This doesn't seem to be used anywhere, maybe remove it?

morantr replied:
Answered on: https://github.com/prplfoundation/prplMesh/pull/1271#discussion_r426739414

arnout reviewed (commented):
Reviewed up to "agent: Make the Agent compile successfully".

I think the first few commits could be squashed together - but don't bother doing that.

This PR is a bit big, so maybe it's possible to split off a few commits that *don't* break stuff into a separate PR? E.g. remove old 4addr should be non-breaking, and the logger changes as well.

morantr committed:
agent: Initial commit

Create fronthaul_manager folder and move there monitor and ap_manager files.
From this commit compilation will be broken, but will be fixed in future
commit.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
cmake: Change add_subdirectory call to "fronthaul_manager"

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agnet: Move monitors cmakelist and main.cpp file to upper directory

Change the monitor main file to fronthaul main file.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
cmake: Modify monitor CMakeList file

Change project name to "beerocks_fronthaul".
Change file locations to add to the fronthaul_manager_sources file.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Change executable name on prplmesh_utils.sh

Since the beerocks_monitor executable is no longer exist and there is a beerocks_fronthaul instead,
change the executable name on prplmesh_utils.sh script.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Make the Agent compile successfully

Moving the ap_manager and the monitor to a new location has broken the Agent compilation.
This commit adds minor changes to make the Agent compile again:
- Remove any ap_manager object reference on the son_slave.
- ap_manager stop/start function implementation was removed and replaces with empty functions.
On future commit, it will be removed completely.
- ap_manager thread error query on socket disconnect event has bee removed since son_slave
has no longer has access to ap_mamanger object. In future commit, the error code will be
handled properly.
- Change some include paths to compile prplMesh successfully.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Rename "monitor" to "fronthaul" on fronthaul main file

As part of converting the monitor process to fronthaul process which will contain the ap_manager
thread and the monitor thread, rename variables and comments on the fronthaul main file
from "monitor" to "fronthaul".

Add to beerocks_defines.h macros that define strings for beerocks fronthaul and ap_manager.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Add support to multiple logger IDs on beerocks_logging

Since the ap_manager thread and the monitor thread will be on the same process, we will
need separate log files for each one.
Currently the beerocks_logging support single logger ID which is "default".
To support custom logger ID add an additional logger_id argument to beerocks_logging
constructor and add logger_id member to the logging class that will be used to configure
the logger instead of "default" logger ID.

On the way, make the module_name on the constructor to be mandatory instead of optional, since every process already passes this argument.
Also, change the logging class string argument to be a const reference.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Initialize ap_manager and monitor logger instances

Add a function that creates a logger instance for a given file name.
Use this function to create separate logger instances for ap_manager and monitor threads.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Modify ap_manager constructor

Add "iface" string and the "logging" class, an arguments to ap_manager constructor.
While at it:
- Change the slave_uds argument to const reference, and move its initialization to
the constructor body.
- Change "bss_steer_valid_int" and "bss_steer_imminent_valid_int" members to be
initialized on the class definition instead of the constructor.
Also, move the default values from beerocks_defines to the ap_manager class since
they are not being used from somewhere else.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Remove 4addr pending station mechanism

Back then, the backhaul manager used to connect to the AP with 2 steps:
connecting in 3addr mode, disconnect, and then connect in 4addr mode.
This flow is no longer exist, and the backhaul manager now connects immediately
on 4addr mode.

Still, there are leftovers of this flow on son_slave and ap_manager which are no longer
needed so they are being removed.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Remove messages related to old 4addr connection flow

Since there will still be changes on the tlvf, auto generated files commit
will be at the end of the PR.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add FSM skeleton to ap_manager

There is no purpose in bringing up monitor thread before the ap_manager
has finished the attachment process.

Add FSM skeleton that will be filled on future commits, to allow getting
the ap_manager state from outside and so, the fronthaul main file could
know when to bring up the monitor thread.

Creating FSM has benefits such as making the code more ordered and readable,
and also this would allow splitting all the code which is now on after_select
function to separate states.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Bring up ap_manager on fronthaul main file

Add creation of ap_manager instance on fronthaul main file and start
the monitor thread only after the ap_manager has attached to BWL.

Since now the ap_manager init function is called from outside, move it
to "public" on the class definition.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add new messages to ap_manager

Add new messages to ap_manager that will be used to change the ap_manages state machine state.
ACTION_APMANAGER_UP_NOTIFICATION will be sent by the ap_manager to the son_slave, and the
son_slave will respond it with ACTION_APMANAGER_CONFIGURE message.

Auto-generated code will be added at the end of the PR.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add connect_to_agent() to ap_manager

Add connect_to_agent() to ap_manager. Moved part of the code from the init function to the connect function.
Call this function on the after_select function and also add a call to the FSM function.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Fill INIT and WAIT_FOR_CONFIGURATION on ap_manager FSM

Fill INIT and WAIT_FOR_CONFIGURATION on ap_manager FSM.
The INIT state send ACTION_APMANAGER_UP_NOTIFICATION to the Agent and
move to WAIT_FOR_CONFIGURATION.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add handler to ACTION_APMANAGER_UP_NOTIFICATION on son_slave

Add a handler to ACTION_APMANAGER_UP_NOTIFICATION on son_slave that responds with
ACTION_APMANAGER_CONFIGURE message.
This handler replaces ACTION_APMANAGER_INIT_DONE_NOTIFICATION handler.

The configuration on the message replaces the configuration which has been configured on
the ap_manager_start function which will be removed on the future.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add ACTION_APMANAGER_CONFIGURE message handler

Add on the ap_manager a handler to ACTION_APMANAGER_CONFIGURE message sent by son_slave.
The handler configures class calls to a newly added function that creates the ap_wlan_hal instance,
instead of where it used to be - init function.
At the end of the handler, move the ap_manager to attaching state.

Remove ap_manager_config() and ap_manager_conf_t struct which are no longer used.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Fill ap_manager ATTACHED, ATTACHED and OPERATIONAL states

Copy the code from after_select function on ap_manager and split it to the above
states on the state machine.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add stop and start function to fronthaul on son_slave

The added function is a copy of the monitor stop and start function with some renaming.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Change son_slave FSM to new ap_manager init sequence

Remove the state that waits for ap_manager join message
Remove the state that starts the monitor process.
Remove the call that starts the ap_mananger thread.
Add a call that starts the fronthaul process.

Also, add some code that protects from a race between ap_manager and monitor join message.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Remove ap_manager and monitor stop and start functions

These functions are no longer needed so they have been removed.
the call for the stop function has been replaced with a call to fronthaul_stop() call.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Remove APMANAGER_INIT_DONE_NOTIFICATION

ACTION_APMANAGER_INIT_DONE_NOTIFICATION is removed since it will no longer be used.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Update auto-generated files

Update auto-generated files since the last change on TLVF has been committed.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Change initialization of socket thread m_select_wake_up_time

Since the monitor object is created but the thread is not starting right away, a lot of time passes
between class initialization to first select call, which causes an error being printed to the log.

Initialize m_select_wake_up_time to min value which turns out to be a negative value and
promises that the error print would not be printed right after the monitor thread starts.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Move sending of ACTION_MONITOR_JOINED_NOTIFICATION

ACTION_MONITOR_JOINED_NOTIFICATION is currently being sent on the init function of the monitor.
This is not making much sense since at the moment the monitor is not attached and cannot do
anything.
Move the sending of the ACTION_MONITOR_JOINED_NOTIFICATION message to right after the monitor
has finished the attachment process.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Add custom logger ID helper file

When using LOG() macro it is always print to default logger ID which is "default" string.
When we want to have to log files on a single process we have to create a different
logger ID for the non-default logger.

We can't use LOG() macro if we want to print to non-default logger ID.
To avoid replacing all LOG() calls manually on some file, add a helper file that allows with
a single #define and #include at the beginning of a file, do the replacing automatically.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Activate custom logging on monitor files

Activate it only on the monitor_thread file and not on other monitor
files since this solution is not a complete one and not covering cases
such as printing on files from a different lib such as bcl, bwl etc...

On the follow-up commit, a complementary solution will be provided:
Option to attach the thread ID to a specific logger ID and in that way,
any print on the thread will be printed to the right file.
Although it will cover most of the cases, still it is not covering prints
on functions that run in the context of another thread such as Monitor
thread constructor and init function.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
easylogging: Add option to attach thread ID to logger ID

To make two threads on the same process prints to different files be fully supported,
add an option on easylogging to attach the current thread ID to the logger ID.
When trying to print, easylogging will check there is any logger ID is attached to the
current thread, and if there is, it will use it.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Wrap option to attach logger ID to thread ID on beerocks logging

Wrap the newly added feature on easylogging from the previous commit, on beerocks logging class.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Attach the Monitor thread ID to the Monitor logger

Use the feature added on the previous commit to attach the Monitor thread ID to its logger ID.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

arnout committed:
tests: environment.py: support multiple log files

As part of the unified agent refactoring, the per-radio logs are now
spread over three files instead of two: agent_wlanX, ap_manager_wlanX
and monitor_wlanX. The agent_wlanX log file will eventually be removed
again, and we should look for that information in ap_manager_wlanX
instead.

For the time being, however, it's easier to simply look in both of them.
When unified agent is done, the cases where we look in agent_wlanX can
instead be converted into check_log on the agent instead of the radio.

Until that time, as a workaround, support two log files on the radio:
agent and ap_manager. Update _docker_wait_for_log to support multiple
programs instead of a single one, and update the callers to pass a list
instead of a single entry.

This change should be reverted again when unified agent is done.
Therefore, add a big comment before _docker_wait_for_log that says so.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

morantr committed:
tools: Add cppcheck suppressions list

Add an option to suppress errors on cppcheck so it will not complain about the errors on the list.
Added to the suppression list two types of errors that are related to class member initialization:

uninitMemberVar - Complains about a member is initialized on the class definition instead
of the class constructor.
IMO it is much more readable to give default value on the class definition rather on the
constructor.

useInitializationList - Complains about initialization of class members on the constructor
body instead of the initializer list.
IMO initializing class member on the initializer list is less readable since they are separated
by commas in one or more long lines instead of initializing them line by line on the
constructor body.
Also, initializing class members on the initializer list must be done on the order they are
defined on the class otherwise compilation will fail. This is very annoying especially when we
want to add a new class member and then we have to change the order of the initializer list.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
cppcheck: Update exist issues file

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Add an option to set thread name on beerocks logging

When using Syslog, easylogging does not write the prints to log files, it sends it to Syslog.
On Syslog, it is possible to distinguish prints source by process name.
So in our case, it will be  "beerocks_fronthaul_wlanX", and we need a way to separate between
ap_manager and monitor threads.
To do it, add an option to set the thread name, on logging class, and set it from ap_manager
and monitor thread.
After the change the log will look like this on Syslog:
beerocks_fronthaul_wlan0: <ap_manager> ap_manager_thread.cpp[205] --> waiting to attach to wlan0
beerocks_fronthaul_wlan0: <monitor> monitor_thread.cpp[205] --> waiting to attach to wlan0

The thread name between <> brackets will be used to split the messages from the main log file
to the right prplMesh log file.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
easylogging: Syslog adaptations

Currently, beerocks logging configure two logger IDs for each instance.
One is the "default" or any given ID name, and the other one is "syslog"
which has a different configuration.
This is not required. We can use only our main logger ID and print to Syslog.
On this commit unify the default logger and the Syslog logger to one by using
the same configuration instance.
If Syslog is configured, just override the main configuration so it would
fit Syslog.

On Easylogging, change the logger ID in case of Syslog is configured, from
Syslog logger to the default logger.

Add logger ID argument to the version function so it would print version
to the right logger.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -1645,10 +1639,31 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
+        }
+
         ap_manager_socket = sd;
-        slave_state       = STATE_WAIT_FOR_AP_MANAGER_JOINED;
+        add_socket(ap_manager_socket);
```
TBH, I think it's silly to go through the listen-accept dance for intra-process communication. It would make a lot more sense to simply create a socketpair() and pass one of them to the child. In fact, the same can be used even if it's a separate process, because the file descriptors are inherited by forked children.

But OK, that's a completely unrelated refactoring.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -1697,12 +1697,13 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
             supported_channels.begin(), &std::get<1>(tuple_supported_channels),
             &std::get<1>(tuple_supported_channels) + notification->supported_channels_size());
 
+        // If Monitor has already joined the stats has been incremented on the Monitor join handler,
```
I don't understand this... Do you mean
```suggestion
        // If Monitor has already joined the state has been incremented on the Monitor join handler,
```
?

morantr replied:
yes, fixed.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -2424,7 +2425,10 @@ bool slave_thread::handle_cmdu_monitor_message(Socket *sd,
+        // Since the ap_manager and the Monitor are brought up together, the monitor join could be
+        // received before the ap_manager join message.
+        if (slave_state == STATE_WAIT_FOR_MONITOR_JOINED ||
+            slave_state == STATE_WAIT_FOR_AP_MANAGER_JOINED) {
```
So now, when we receive the monitor join but no apmanager join, we directly go to backhaul enable state... That doesn't sound right...

I think the only way to start them in parallel is to have a single state "WAIT_FOR_THREADS_JOINED" and a separate boolean for each of them. You don't even need a boolean, just check monitor_socket and ap_manager_socket.

morantr replied:
Done.

arnout commented on common/beerocks/bcl/include/bcl/beerocks_socket_thread.h:
```diff
@@ -97,9 +97,10 @@ class socket_thread : public thread_base {
+    uint32_t m_select_timeout_msec  = 0;
+    bool m_skip_next_select_timeout = false;
+    std::chrono::steady_clock::time_point m_select_wake_up_time =
+        std::chrono::steady_clock::time_point::min();
```
I think this would be better:
```suggestion
        std::chrono::steady_clock::time_point::max();
```
Indeed, what you want to avoid is for this condition to trigger:
```
    auto awake_time_msec = std::chrono::duration_cast<std::chrono::milliseconds>(
                               std::chrono::steady_clock::now() - m_select_wake_up_time)
                               .count();

    if (m_select_timeout_msec > 0 && awake_time_msec > m_select_timeout_msec)
```
When you initialize it with min(), awake_time_msec should actually be very large (and positive, since now is definitely later than min) so larger than m_select_timeout_msec. The only reason that it happens to work is that it is so large that it overflows and thus becomes negative. But you can hardly rely on that... With max, it simply be negative and the condition won't trigger.

(My analysis could be wrong though.)

morantr replied:
Apparently both give the same value: https://repl.it/repls/BriefPoorKeygen
But no problem, I'll change it to max.

arnout commented on common/beerocks/bcl/include/bcl/beerocks_logging_custom.h:
```diff
@@ -0,0 +1,33 @@
+ * See LICENSE file for more details.
+ */
+
+/**
+ * @brief This file helps to define and use a custom logger id which is required for using
+ * easylogging with more than one log files on a single thread.
+ * 
+ * To use custom logger id:
+ * 1. Create beerock::logging object with custom logging id string: "logging_id_str"
+ * 2. Define BEEROCKS_CUSTOM_LOGGER_ID macro as logging_id_str":
+ *    #define BEEROCKS_CUSTOM_LOGGER_ID "logging_id_str"
+ * 3. Include this file where there is a need to write to the non-default logger id.
+ */
```
I could be wrong, but isn't this exactly the same as defining `ELPP_DEFAULT_LOGGER` before including easylogging++.h?

morantr replied:
Now that you say that and I look at that, it seems that it will give us the same result.
I didn't know about this option...
I will test it, and it will work, I'll remove this commit, and fix the follow-up commit.
I'm just not sure if it will affect other files on another thread.

arnout replied:
ELPP_DEFAULT_LOGGER is only used to set kDefaultLoggerId which is a module-local variable. So it doesn't help for functions from libraries. But it can be used for the "main" functions.

morantr replied:
@arnout 
I tested it with:
```cpp
#define BEEROCKS_CUSTOM_LOGGER_ID BEEROCKS_MONITOR
#include <bcl/beerocks_logging_custom.h>
```
It doesn't work. I see the monitor constructor print on the ap_manager.
I guess it is because the ap_manager is the default logger.
Anyway, I leave it as I wrote it.

arnout commented on agent/src/beerocks/fronthaul_manager/monitor/monitor_thread.cpp:
```diff
@@ -10,7 +10,9 @@
 #include <bcl/beerocks_utils.h>
 #include <bcl/network/network_utils.h>
-#include <easylogging++.h>
+
+#define BEEROCKS_CUSTOM_LOGGER_ID BEEROCKS_MONITOR
+#include <bcl/beerocks_logging_custom.h>
```
As noted elsewhere:
```suggestion

#define ELPP_DEFAULT_LOGGER BEEROCKS_MONITOR
#include <easylogging++.h>
```

Also, just to be sure that easylogging++.h is not included earlier in any of the already-included header files, move the definition before *all* includes.

arnout commented on agent/src/beerocks/fronthaul_manager/monitor/monitor_db.cpp:
```diff
@@ -10,7 +10,6 @@
 
 #include <bcl/network/network_utils.h>
 #include <bcl/son/son_wireless_utils.h>
-#include <easylogging++.h>
```
This *happens* to work because network_utils.h includes easylogging++.h, but it's not good to rely on that. So please don't remove this.

morantr replied:
returned it.

arnout commented on framework/external/easylogging/easylogging++.cc:
```diff
@@ -2605,7 +2605,12 @@ Writer& Writer::construct(int count, const char* loggerIds, ...) {
 void Writer::initializeLogger(const std::string& loggerId, bool lookup, bool needLock) {
   if (lookup) {
-    m_logger = ELPP->registeredLoggers()->get(loggerId, ELPP->hasFlag(LoggingFlag::CreateLoggerAutomatically));
+    auto thread_logger_id = ELPP->getThreadLoggerId();
```
I think it would be more appropriate to use the thread logger ID as the equivalent of a default - now the thread logger is taking precedence over an explicitly specified one, which I don't think is appropriate. In other words, define ELPP_CURR_FILE_LOGGER_ID as ELPP->getThreadLoggerId(). However, that would have far-reaching effects, and at least for us it's not necessary, so let's keep it as is.

morantr replied:
Yes, I prefer doing the minimal possible change on easylogging.

arnout commented on tools/docker/static-analysis/suppressions.txt:
```diff
@@ -0,0 +1,5 @@
+// Suppress warning: Member variable 'X' is not initialized in the constructor.
+uninitMemberVar
```
This warning should not be given for a member that is initialized in the struct... For example, in ap_manager_thread, the warning is given for acs_enabled (which is not initialized in the class definition nor in the constructor) but not for ap_hal_ext_events (which is initialized in the class definition but not in the constructor).

So IMO this suppression should not be used... Do you have a concrete example where it goes wrong?

morantr replied:
If I remember correctly, I say this warning when a class member was initialized on class definition instead of on the initialized list.
I will recheck it.

arnout commented on tools/docker/static-analysis/suppressions.txt:
```diff
@@ -0,0 +1,5 @@
+// Suppress warning: Member variable 'X' is not initialized in the constructor.
+uninitMemberVar
+
+// Suppress performance: Variable 'X' is assigned in constructor body. Consider performing initialization in initialization list.
+useInitializationList
```
This one makes sense - since we diverge from the recommendation, it's clear that we need to suppress the warnings for it.

arnout commented on agent/src/beerocks/fronthaul_manager/fronthaul_manager_main.cpp:
```diff
@@ -178,15 +178,19 @@ int main(int argc, char *argv[])
     // Init logger ap_manager
-    std::string base_ap_manager_name = std::string(BEEROCKS_AP_MANAGER) + "_" + fronthaul_iface;
+    std::string base_ap_manager_name =
+        std::string(syslog ? BEEROCKS_FRONTHAUL : BEEROCKS_AP_MANAGER) + "_" + fronthaul_iface;
```
This smells like a serious break of abstraction... Could this condition somehow be moved into init_logger itself?

morantr replied:
No, I wish...
init_logger is just a static function I created in this file. I still would have to provide the two options so it will be kind of the same.

arnout commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.h:
```diff
@@ -87,10 +87,7 @@ class ap_manager_thread : public beerocks::socket_thread {
-    uint8_t wifi_channel;
-    uint8_t m_iface_type;
     bool acs_enabled;
-    bool low_filter;
```
This shouldn't be a fixup commit but a real separate commit.

morantr replied:
You will forgive me for not adding more commits to this PR, right? :)

arnout reviewed (approved):
I don't want to look at this any more so I'm approving even though I have so many comments :-)

vitalybu commented on agent/src/beerocks/fronthaul_manager/monitor/monitor_thread.cpp:
```diff
@@ -213,24 +215,19 @@ bool monitor_thread::init()
 
 void monitor_thread::after_select(bool timeout)
 {
+    if (!m_logger_configured) {
```
Please move this to `before_select`

morantr replied:
Why? Currently, we don't override "before_select" on this thread.

arnout replied:
> Why?

Because it makes sure that any logs that are done during the first select are also done correctly.

Ideally, it should be done in a thread initialization function, but beerocks_thread_base doesn't have that (init() is called from the context of the parent, not of the thread itself).

> Currently, we don't override "before_select" on this thread.

That's not a good reason :-)

morantr replied:
Ok added "before_select" and move it to there.

vitalybu commented on agent/src/beerocks/fronthaul_manager/fronthaul_manager_main.cpp:
```diff
@@ -147,60 +178,82 @@ int main(int argc, char *argv[])
+
+    if (!ap_manager.init()) {
+        CLOG(ERROR, g_logger_ap_mananger->get_logger_id()) << "ap manager init() has failed!";
+        return 0;
```
That's an error, so you should probably return a non-zero value.
Same goes for every other error returned by this module.

morantr replied:
Done, but changed only this return value.
This PR is big enough and I will need to do a separate commit for this.
So probably it will be better to change it here and on any other process as part of a different PR.

vitalybu commented on agent/src/beerocks/fronthaul_manager/fronthaul_manager_main.cpp:
```diff
@@ -147,60 +178,82 @@ int main(int argc, char *argv[])
+            if (monitor.is_running()) {
+                continue;
+            } else if (!monitor.start()) {
+                CLOG(ERROR, g_logger_monitor->get_logger_id()) << "monitor.init() has failed";
```
```suggestion
                CLOG(ERROR, g_logger_monitor->get_logger_id()) << "monitor.start() has failed";
```

morantr replied:
fixed.

vitalybu commented on agent/src/beerocks/fronthaul_manager/fronthaul_manager_main.cpp:
```diff
@@ -147,60 +178,82 @@ int main(int argc, char *argv[])
+        // The ap_manager is the main process thread. After the ap_manager is finished the attach
+        // process, start the monitor thread. There is no point to start it before.
+        auto ap_manager_state = ap_manager.get_state();
+        if (ap_manager_state < son::ap_manager_thread::eApManagerState::ATTACHED) {
```
This condition is not really needed.
It's enough to if the AP Manager state is `>= ATTACHED` and the monitor is not running...

morantr replied:
Yes, you right. Remove it and leave only the condition on the "else".

vitalybu commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -1669,10 +1637,29 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
+    case beerocks_message::ACTION_APMANAGER_UP_NOTIFICATION: {
+        LOG(INFO) << "received ACTION_APMANAGER_UP_NOTIFICATION from sd=" << intptr_t(sd);
+        if (ap_manager_socket) {
+            LOG(ERROR) << "AP manager opened new socket altough the exist socket is already open!";
```
```suggestion
            LOG(ERROR) << "AP manager opened new socket altough the existing socket is still open!";
```

vitalybu replied:
Also, not sure it's an ERROR (maybe a WARNING), as we simply close the old socket and continue...

morantr replied:
The thing is, that is should not occur at all, that is why it is an error.
Changed the print.

vitalybu commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3760,34 +3687,39 @@ bool slave_thread::hostap_services_on()
-    if (did_stop)
-        LOG(DEBUG) << "monitor_stop() - done";
+    // kill monitor slave pid
+    os_utils::kill_pid(config.temp_path,
```
Isn't the monitor is just a thread in our process now? What are we trying to kill here?

morantr replied:
It kills the fronthaul process. The comments are wrong. Fix them.

vitalybu commented on common/beerocks/bcl/source/beerocks_logging.cpp:
```diff
@@ -506,7 +529,13 @@ void logging::apply_settings()
+    auto logger = el::Loggers::getLogger(m_logger_id);
+    if (!logger) {
+        LOG(ERROR) << "invalid logger!";
+        return;
```
If this function can now fail, it should return a (boolean?) value...

morantr replied:
It could have failed before, I've just moved it.
Anyway, I'm not really sure what to do with that and I don't want to do an additional refactoring as part of this PR.

vitalybu commented on framework/external/easylogging/easylogging++.h:
```diff
@@ -3688,6 +3703,14 @@ class Helpers : base::StaticClass {
   static inline void setArgs(int argc, const char** argv) {
     ELPP->setApplicationArguments(argc, const_cast<char**>(argv));
   }
+
```
Remove extra newline

morantr replied:
done.

kantera800 reviewed (commented):
This is a big change so please don't merge this until RDKB/UGW platforms will pass sanity,
Issues as #1404 should be resolved first

morantr committed:
common: Change print condition

On system_call function, we print error message if the system command returns non zero.
But system command returns 0 on success and non-zero on failure.
Change the condition for the print so it would print an error message on a non-zero return value.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

rmelotte commented:
I'm also to blame because I didn't review it, but this should never have been merged without at the very least running /some/ of the agent certification tests (using the magic string to also make it run after mergify's rebases).
@morantr  please take it into account for the next parts of https://github.com/prplfoundation/prplMesh/issues/1082

morantr commented:
> I'm also to blame because I didn't review it, but this should never have been merged without at the very least running /some/ of the agent certification tests (using the magic string to also make it run after mergify's rebases).
> @morantr please take it into account for the next parts of #1082

Since I didn't change anything related to certification flows, it was not required.
All the things I changed has been tested.
The future PRs will actually change things related to the certification flows so it will be tested.
Anyway, as we debug the issue it seems that the root cause is not my change.
@rmelotte
